File: clients/client-go/ack_test.go
package queen

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

// The server (routes/ack.cpp + queen.ack_messages_v2) responds to both
// /api/v1/ack and /api/v1/ack/batch with a top-level JSON array, one item per
// acknowledgment in request order:
//
//	[{"index":0,"transactionId":"...","success":false,"error":"Invalid or expired lease",
//	  "queueName":"q","partitionName":"Default","leaseReleased":false,"dlq":false}]
//
// HttpClient wraps top-level arrays as {"data": [...]}. These tests pin that
// wire format so a rejected ack/nack is never reported as Success=true.

func newAckTestServer(t *testing.T, wantPath string, respond func(body map[string]interface{}) interface{}) *Queen {
	t.Helper()
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != wantPath {
			t.Errorf("unexpected request path %q, want %q", r.URL.Path, wantPath)
			http.NotFound(w, r)
			return
		}
		var body map[string]interface{}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Errorf("failed to decode request body: %v", err)
		}
		w.Header().Set("Content-Type", "application/json")
		if err := json.NewEncoder(w).Encode(respond(body)); err != nil {
			t.Errorf("failed to encode response: %v", err)
		}
	}))
	t.Cleanup(server.Close)

	client, err := New(server.URL)
	if err != nil {
		t.Fatalf("failed to create client: %v", err)
	}
	t.Cleanup(func() { client.Close(context.Background()) })
	return client
}

func ackResultItem(index int, txnID string, success bool, errMsg interface{}) map[string]interface{} {
	return map[string]interface{}{
		"index":         index,
		"transactionId": txnID,
		"success":       success,
		"error":         errMsg,
		"queueName":     "test-queue",
		"partitionName": "Default",
		"leaseReleased": success,
		"dlq":           false,
	}
}

func TestAckSingleServerResponse(t *testing.T) {
	cases := []struct {
		name        string
		ackSuccess  bool // the status the client sends (true=completed, false=failed)
		serverItem  map[string]interface{}
		wantSuccess bool
		wantError   string
	}{
		{
			name:        "accepted ack",
			ackSuccess:  true,
			serverItem:  ackResultItem(0, "tx-1", true, nil),
			wantSuccess: true,
		},
		{
			name:        "ack rejected for expired lease",
			ackSuccess:  true,
			serverItem:  ackResultItem(0, "tx-1", false, "Invalid or expired lease"),
			wantSuccess: false,
			wantError:   "Invalid or expired lease",
		},
		{
			name:        "nack rejected for missing message",
			ackSuccess:  false,
			serverItem:  ackResultItem(0, "tx-1", false, "Message not found"),
			wantSuccess: false,
			wantError:   "Message not found",
		},
		{
			name:        "rejected without error string still fails",
			ackSuccess:  true,
			serverItem:  ackResultItem(0, "tx-1", false, nil),
			wantSuccess: false,
			wantError:   "acknowledgment rejected by server",
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			wantStatus := AckStatusCompleted
			if !tc.ackSuccess {
				wantStatus = AckStatusFailed
			}
			client := newAckTestServer(t, "/api/v1/ack", func(body map[string]interface{}) interface{} {
				if got := body["status"]; got != wantStatus {
					t.Errorf("request status = %v, want %q", got, wantStatus)
				}
				return []interface{}{tc.serverItem}
			})

			msg := &Message{TransactionID: "tx-1", PartitionID: "p-1", LeaseID: "lease-1"}
			responses, err := client.Ack(context.Background(), msg, tc.ackSuccess, AckOptions{})
			if err != nil {
				t.Fatalf("Ack returned error: %v", err)
			}
			if len(responses) != 1 {
				t.Fatalf("got %d responses, want 1", len(responses))
			}
			if responses[0].Success != tc.wantSuccess {
				t.Errorf("Success = %v, want %v", responses[0].Success, tc.wantSuccess)
			}
			if responses[0].Error != tc.wantError {
				t.Errorf("Error = %q, want %q", responses[0].Error, tc.wantError)
			}
		})
	}
}

func TestAckBatchMixedResults(t *testing.T) {
	client := newAckTestServer(t, "/api/v1/ack/batch", func(body map[string]interface{}) interface{} {
		acks, _ := body["acknowledgments"].([]interface{})
		if len(acks) != 2 {
			t.Errorf("request has %d acknowledgments, want 2", len(acks))
		}
		return []interface{}{
			ackResultItem(0, "tx-1", true, nil),
			ackResultItem(1, "tx-2", false, "Invalid or expired lease"),
		}
	})

	msgs := []*Message{
		{TransactionID: "tx-1", PartitionID: "p-1", LeaseID: "lease-1"},
		{TransactionID: "tx-2", PartitionID: "p-1", LeaseID: "lease-1"},
	}
	responses, err := client.Ack(context.Background(), msgs, true, AckOptions{})
	if err != nil {
		t.Fatalf("Ack returned error: %v", err)
	}
	if len(responses) != 2 {
		t.Fatalf("got %d responses, want 2", len(responses))
	}
	if !responses[0].Success || responses[0].Error != "" {
		t.Errorf("responses[0] = %+v, want success", responses[0])
	}
	if responses[1].Success || responses[1].Error != "Invalid or expired lease" {
		t.Errorf("responses[1] = %+v, want rejected with lease error", responses[1])
	}
}

func TestAckBatchCountMismatch(t *testing.T) {
	client := newAckTestServer(t, "/api/v1/ack/batch", func(body map[string]interface{}) interface{} {
		return []interface{}{ackResultItem(0, "tx-1", true, nil)}
	})

	msgs := []*Message{
		{TransactionID: "tx-1", PartitionID: "p-1", LeaseID: "lease-1"},
		{TransactionID: "tx-2", PartitionID: "p-1", LeaseID: "lease-1"},
	}
	if _, err := client.Ack(context.Background(), msgs, true, AckOptions{}); err == nil {
		t.Fatal("Ack should fail when the server returns fewer results than messages sent")
	}
}

func TestAckErrorEnvelope(t *testing.T) {
	client := newAckTestServer(t, "/api/v1/ack", func(body map[string]interface{}) interface{} {
		return map[string]interface{}{"error": "internal error"}
	})

	msg := &Message{TransactionID: "tx-1", PartitionID: "p-1", LeaseID: "lease-1"}
	responses, err := client.Ack(context.Background(), msg, true, AckOptions{})
	if err != nil {
		t.Fatalf("Ack returned error: %v", err)
	}
	if len(responses) != 1 || responses[0].Success {
		t.Fatalf("responses = %+v, want single rejected response", responses)
	}
	if responses[0].Error != "internal error" {
		t.Errorf("Error = %q, want %q", responses[0].Error, "internal error")
	}
}

func TestAckUnexpectedResponseFormat(t *testing.T) {
	client := newAckTestServer(t, "/api/v1/ack", func(body map[string]interface{}) interface{} {
		return map[string]interface{}{"ok": true}
	})

	msg := &Message{TransactionID: "tx-1", PartitionID: "p-1", LeaseID: "lease-1"}
	if _, err := client.Ack(context.Background(), msg, true, AckOptions{}); err == nil {
		t.Fatal("Ack should fail on a response that is neither a result array nor an error envelope")
	}
}

File: clients/client-js/test-v2/streams-unit/ack.test.js
/**
 * Queen.ack() wire-format tests.
 *
 * The server (routes/ack.cpp + queen.ack_messages_v2) responds to both
 * /api/v1/ack and /api/v1/ack/batch with a top-level JSON array, one item per
 * acknowledgment in request order:
 *
 *   [{"index":0,"transactionId":"...","success":false,"error":"Invalid or expired lease",
 *     "queueName":"q","partitionName":"Default","leaseReleased":false,"dlq":false}]
 *
 * These tests pin that wire format so a rejected ack/nack (e.g. "Invalid or
 * expired lease", "Message not found") is never reported as success: true.
 * Mirror of clients/client-go/ack_test.go.
 */

import { describe, it } from 'node:test'
import assert from 'node:assert/strict'
import { createServer } from 'node:http'

import { Queen } from '../../client-v2/index.js'

function ackResultItem(index, transactionId, success, error = null) {
  return {
    index,
    transactionId,
    success,
    error,
    queueName: 'test-queue',
    partitionName: 'Default',
    leaseReleased: success,
    dlq: false
  }
}

/**
 * Spin up a real HTTP server, point a Queen client at it, run the test body,
 * then tear both down. respond(body) returns the JSON payload to send back.
 * Requests are recorded as { path, body } for assertions.
 */
async function withAckServer(respond, run) {
  const requests = []
  const server = createServer((req, res) => {
    let raw = ''
    req.on('data', chunk => { raw += chunk })
    req.on('end', () => {
      const body = raw ? JSON.parse(raw) : {}
      requests.push({ path: req.url, body })
      res.writeHead(200, { 'Content-Type': 'application/json' })
      res.end(JSON.stringify(respond(body)))
    })
  })
  await new Promise(resolve => server.listen(0, '127.0.0.1', resolve))
  const { port } = server.address()

  const client = new Queen({
    url: `http://127.0.0.1:${port}`,
    handleSignals: false,
    retryAttempts: 1
  })

  try {
    await run(client, requests)
  } finally {
    await client.close()
    await new Promise(resolve => server.close(resolve))
  }
}

const MSG = { transactionId: 'tx-1', partitionId: 'p-1', leaseId: 'lease-1' }

describe('Queen.ack single (/api/v1/ack)', () => {
  it('reports success for an accepted ack', async () => {
    await withAckServer(
      () => [ackResultItem(0, 'tx-1', true)],
      async (client, requests) => {
        const result = await client.ack(MSG, true)
        assert.equal(result.success, true)
        assert.equal(result.error, null)
        assert.equal(result.transactionId, 'tx-1')
        assert.equal(requests[0].path, '/api/v1/ack')
        assert.equal(requests[0].body.status, 'completed')
      }
    )
  })

  it('reports failure when the ack is rejected for an expired lease', async () => {
    await withAckServer(
      () => [ackResultItem(0, 'tx-1', false, 'Invalid or expired lease')],
      async (client) => {
        const result = await client.ack(MSG, true)
        assert.equal(result.success, false)
        assert.equal(result.error, 'Invalid or expired lease')
      }
    )
  })

  it('reports failure when a nack is rejected for a missing message', async () => {
    await withAckServer(
      () => [ackResultItem(0, 'tx-1', false, 'Message not found')],
      async (client, requests) => {
        const result = await client.ack(MSG, false)
        assert.equal(result.success, false)
        assert.equal(result.error, 'Message not found')
        assert.equal(requests[0].body.status, 'failed')
      }
    )
  })

  it('still fails when rejected without an error string', async () => {
    await withAckServer(
      () => [ackResultItem(0, 'tx-1', false, null)],
      async (client) => {
        const result = await client.ack(MSG, true)
        assert.equal(result.success, false)
        assert.equal(result.error, 'Acknowledgment rejected by server')
      }
    )
  })

  it('propagates a top-level error envelope', async () => {
    await withAckServer(
      () => ({ error: 'internal error' }),
      async (client) => {
        const result = await client.ack(MSG, true)
        assert.equal(result.success, false)
        assert.equal(result.error, 'internal error')
      }
    )
  })

  it('fails on a response that is neither a result array nor an error envelope', async () => {
    await withAckServer(
      () => ({ ok: true }),
      async (client) => {
        const result = await client.ack(MSG, true)
        assert.equal(result.success, false)
        assert.match(result.error, /Unexpected ack response format/)
      }
    )
  })
})

describe('Queen.ack batch (/api/v1/ack/batch)', () => {
  const MSGS = [
    { transactionId: 'tx-1', partitionId: 'p-1', leaseId: 'lease-1' },
    { transactionId: 'tx-2', partitionId: 'p-1', leaseId: 'lease-1' }
  ]

  it('reports success when every acknowledgment is accepted', async () => {
    await withAckServer(
      () => [ackResultItem(0, 'tx-1', true), ackResultItem(1, 'tx-2', true)],
      async (client, requests) => {
        const result = await client.ack(MSGS, true)
        assert.equal(result.success, true)
        assert.equal(result.processed, 2)
        assert.equal(result.results.length, 2)
        assert.equal(requests[0].path, '/api/v1/ack/batch')
        assert.equal(requests[0].body.acknowledgments.length, 2)
      }
    )
  })

  it('reports failure with per-item results on mixed outcomes', async () => {
    await withAckServer(
      () => [
        ackResultItem(0, 'tx-1', true),
        ackResultItem(1, 'tx-2', false, 'Invalid or expired lease')
      ],
      async (client) => {
        const result = await client.ack(MSGS, true)
        assert.equal(result.success, false)
        assert.equal(result.error, 'Invalid or expired lease')
        assert.equal(result.results[0].success, true)
        assert.equal(result.results[1].success, false)
        assert.equal(result.results[1].error, 'Invalid or expired lease')
      }
    )
  })

  it('fails when the server returns fewer results than acknowledgments sent', async () => {
    await withAckServer(
      () => [ackResultItem(0, 'tx-1', true)],
      async (client) => {
        const result = await client.ack(MSGS, true)
        assert.equal(result.success, false)
        assert.match(result.error, /1 results, expected 2/)
      }
    )
  })

  it('propagates a top-level error envelope to every acknowledgment', async () => {
    await withAckServer(
      () => ({ error: 'internal error' }),
      async (client) => {
        const result = await client.ack(MSGS, true)
        assert.equal(result.success, false)
        assert.equal(result.error, '2 of 2 acknowledgments rejected: internal error')
        assert.equal(result.results.length, 2)
        assert.equal(result.results[0].error, 'internal error')
      }
    )
  })
})
